File: platform/polyglot/confluence_api/mjs/src/services/space-permission-service.mjs
/**
 * @module services/space-permission-service
 * @description Service for Confluence Space Permission REST API operations.
 *
 * Manages space-level permissions including viewing, granting, and revoking
 * permissions for anonymous users, groups, and individual users.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence space permission operations.
 */
export class SpacePermissionService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  /**
   * Get all permissions for a space.
   *
   * @param {string} spaceKey - The space key.
   * @returns {Promise<Object>} Space permissions.
   */
  async getPermissions(spaceKey) {
    log.debug('getPermissions called', { spaceKey });
    try {
      const result = await this._client.get(`space/${spaceKey}/permission`);
      log.info('getPermissions succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('getPermissions failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Add a permission to a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {Object} data - Permission payload (subject, operation).
   * @returns {Promise<Object>} The added permission.
   */
  async addPermission(spaceKey, data) {
    log.debug('addPermission called', { spaceKey });
    try {
      const result = await this._client.post(`space/${spaceKey}/permission`, data);
      log.info('addPermission succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('addPermission failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get anonymous permissions for a space.
   *
   * @param {string} spaceKey - The space key.
   * @returns {Promise<Object>} Anonymous space permissions.
   */
  async getAnonymousPermissions(spaceKey) {
    log.debug('getAnonymousPermissions called', { spaceKey });
    try {
      const result = await this._client.get(`space/${spaceKey}/permission/anonymous`);
      log.info('getAnonymousPermissions succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('getAnonymousPermissions failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get permissions for a specific group in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} groupName - The group name.
   * @returns {Promise<Object>} Group space permissions.
   */
  async getGroupPermissions(spaceKey, groupName) {
    log.debug('getGroupPermissions called', { spaceKey, groupName });
    try {
      const result = await this._client.get(`space/${spaceKey}/permission/group/${groupName}`);
      log.info('getGroupPermissions succeeded', { spaceKey, groupName });
      return result;
    } catch (error) {
      log.error('getGroupPermissions failed', { spaceKey, groupName, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get permissions for a specific user in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} userKey - The user key.
   * @returns {Promise<Object>} User space permissions.
   */
  async getUserPermissions(spaceKey, userKey) {
    log.debug('getUserPermissions called', { spaceKey, userKey });
    try {
      const result = await this._client.get(`space/${spaceKey}/permission/user/${userKey}`);
      log.info('getUserPermissions succeeded', { spaceKey, userKey });
      return result;
    } catch (error) {
      log.error('getUserPermissions failed', { spaceKey, userKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Grant a permission to anonymous users in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {Object} data - Permission grant payload (operation).
   * @returns {Promise<Object>} The granted permission.
   */
  async grantAnonymous(spaceKey, data) {
    log.debug('grantAnonymous called', { spaceKey });
    try {
      const result = await this._client.post(`space/${spaceKey}/permission/anonymous`, data);
      log.info('grantAnonymous succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('grantAnonymous failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Grant a permission to a group in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} groupName - The group name.
   * @param {Object} data - Permission grant payload (operation).
   * @returns {Promise<Object>} The granted permission.
   */
  async grantGroup(spaceKey, groupName, data) {
    log.debug('grantGroup called', { spaceKey, groupName });
    try {
      const result = await this._client.post(`space/${spaceKey}/permission/group/${groupName}`, data);
      log.info('grantGroup succeeded', { spaceKey, groupName });
      return result;
    } catch (error) {
      log.error('grantGroup failed', { spaceKey, groupName, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Grant a permission to a user in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} userKey - The user key.
   * @param {Object} data - Permission grant payload (operation).
   * @returns {Promise<Object>} The granted permission.
   */
  async grantUser(spaceKey, userKey, data) {
    log.debug('grantUser called', { spaceKey, userKey });
    try {
      const result = await this._client.post(`space/${spaceKey}/permission/user/${userKey}`, data);
      log.info('grantUser succeeded', { spaceKey, userKey });
      return result;
    } catch (error) {
      log.error('grantUser failed', { spaceKey, userKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Revoke a permission from anonymous users in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {Object} data - Permission revocation payload (operation).
   * @returns {Promise<void>}
   */
  async revokeAnonymous(spaceKey, data) {
    log.debug('revokeAnonymous called', { spaceKey });
    try {
      const result = await this._client.delete(`space/${spaceKey}/permission/anonymous`, {
        queryParams: data,
      });
      log.info('revokeAnonymous succeeded', { spaceKey });
      return result;
    } catch (error) {
      log.error('revokeAnonymous failed', { spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Revoke a permission from a group in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} groupName - The group name.
   * @param {Object} data - Permission revocation payload (operation).
   * @returns {Promise<void>}
   */
  async revokeGroup(spaceKey, groupName, data) {
    log.debug('revokeGroup called', { spaceKey, groupName });
    try {
      const result = await this._client.delete(`space/${spaceKey}/permission/group/${groupName}`, {
        queryParams: data,
      });
      log.info('revokeGroup succeeded', { spaceKey, groupName });
      return result;
    } catch (error) {
      log.error('revokeGroup failed', { spaceKey, groupName, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Revoke a permission from a user in a space.
   *
   * @param {string} spaceKey - The space key.
   * @param {string} userKey - The user key.
   * @param {Object} data - Permission revocation payload (operation).
   * @returns {Promise<void>}
   */
  async revokeUser(spaceKey, userKey, data) {
    log.debug('revokeUser called', { spaceKey, userKey });
    try {
      const result = await this._client.delete(`space/${spaceKey}/permission/user/${userKey}`, {
        queryParams: data,
      });
      log.info('revokeUser succeeded', { spaceKey, userKey });
      return result;
    } catch (error) {
      log.error('revokeUser failed', { spaceKey, userKey, message: error.message, status: error.status });
      throw error;
    }
  }
}

File: platform/polyglot/confluence_api/py/src/confluence_api/services/space_permission_service.py
"""
Space permission service for Confluence Data Center REST API v9.2.3.

Provides operations for managing space permissions including anonymous,
group-based, and user-based permission grants and revocations.
"""

from __future__ import annotations

from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence_api", __file__)


class SpacePermissionService:
    """Service for Confluence space permission operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    def get_permissions(self, space_key: str) -> dict:
        """Retrieve all permissions for a space."""
        log.debug("get_permissions called", {"space_key": space_key})
        result = self._client.get(f"space/{space_key}/permission")
        log.info("get_permissions succeeded", {"space_key": space_key})
        return result

    def add_permission(self, space_key: str, data: dict) -> dict:
        """Add a permission to a space."""
        log.debug("add_permission called", {"space_key": space_key})
        result = self._client.post(f"space/{space_key}/permission", json_data=data)
        log.info("add_permission succeeded", {"space_key": space_key})
        return result

    def get_anonymous_permissions(self, space_key: str) -> dict:
        """Retrieve anonymous permissions for a space."""
        log.debug("get_anonymous_permissions called", {"space_key": space_key})
        result = self._client.get(f"space/{space_key}/permission/anonymous")
        log.info("get_anonymous_permissions succeeded", {"space_key": space_key})
        return result

    def get_group_permissions(self, space_key: str, group_name: str) -> dict:
        """Retrieve permissions for a specific group in a space."""
        log.debug("get_group_permissions called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = self._client.get(
            f"space/{space_key}/permission/group/{group_name}"
        )
        log.info("get_group_permissions succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    def get_user_permissions(self, space_key: str, user_key: str) -> dict:
        """Retrieve permissions for a specific user in a space."""
        log.debug("get_user_permissions called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = self._client.get(
            f"space/{space_key}/permission/user/{user_key}"
        )
        log.info("get_user_permissions succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result

    def grant_anonymous(self, space_key: str, data: dict) -> dict:
        """Grant anonymous permissions for a space."""
        log.debug("grant_anonymous called", {"space_key": space_key})
        result = self._client.post(
            f"space/{space_key}/permission/anonymous", json_data=data
        )
        log.info("grant_anonymous succeeded", {"space_key": space_key})
        return result

    def grant_group(self, space_key: str, group_name: str, data: dict) -> dict:
        """Grant permissions for a group in a space."""
        log.debug("grant_group called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = self._client.post(
            f"space/{space_key}/permission/group/{group_name}", json_data=data
        )
        log.info("grant_group succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    def grant_user(self, space_key: str, user_key: str, data: dict) -> dict:
        """Grant permissions for a user in a space."""
        log.debug("grant_user called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = self._client.post(
            f"space/{space_key}/permission/user/{user_key}", json_data=data
        )
        log.info("grant_user succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result

    def revoke_anonymous(self, space_key: str, data: dict) -> dict:
        """Revoke anonymous permissions for a space."""
        log.debug("revoke_anonymous called", {"space_key": space_key})
        result = self._client.delete(
            f"space/{space_key}/permission/anonymous",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_anonymous succeeded", {"space_key": space_key})
        return result

    def revoke_group(self, space_key: str, group_name: str, data: dict) -> dict:
        """Revoke permissions for a group in a space."""
        log.debug("revoke_group called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = self._client.delete(
            f"space/{space_key}/permission/group/{group_name}",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_group succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    def revoke_user(self, space_key: str, user_key: str, data: dict) -> dict:
        """Revoke permissions for a user in a space."""
        log.debug("revoke_user called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = self._client.delete(
            f"space/{space_key}/permission/user/{user_key}",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_user succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result


class AsyncSpacePermissionService:
    """Async service for Confluence space permission operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def get_permissions(self, space_key: str) -> dict:
        """Retrieve all permissions for a space."""
        log.debug("get_permissions called", {"space_key": space_key})
        result = await self._client.get(f"space/{space_key}/permission")
        log.info("get_permissions succeeded", {"space_key": space_key})
        return result

    async def add_permission(self, space_key: str, data: dict) -> dict:
        """Add a permission to a space."""
        log.debug("add_permission called", {"space_key": space_key})
        result = await self._client.post(f"space/{space_key}/permission", json_data=data)
        log.info("add_permission succeeded", {"space_key": space_key})
        return result

    async def get_anonymous_permissions(self, space_key: str) -> dict:
        """Retrieve anonymous permissions for a space."""
        log.debug("get_anonymous_permissions called", {"space_key": space_key})
        result = await self._client.get(f"space/{space_key}/permission/anonymous")
        log.info("get_anonymous_permissions succeeded", {"space_key": space_key})
        return result

    async def get_group_permissions(self, space_key: str, group_name: str) -> dict:
        """Retrieve permissions for a specific group in a space."""
        log.debug("get_group_permissions called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = await self._client.get(
            f"space/{space_key}/permission/group/{group_name}"
        )
        log.info("get_group_permissions succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    async def get_user_permissions(self, space_key: str, user_key: str) -> dict:
        """Retrieve permissions for a specific user in a space."""
        log.debug("get_user_permissions called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = await self._client.get(
            f"space/{space_key}/permission/user/{user_key}"
        )
        log.info("get_user_permissions succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result

    async def grant_anonymous(self, space_key: str, data: dict) -> dict:
        """Grant anonymous permissions for a space."""
        log.debug("grant_anonymous called", {"space_key": space_key})
        result = await self._client.post(
            f"space/{space_key}/permission/anonymous", json_data=data
        )
        log.info("grant_anonymous succeeded", {"space_key": space_key})
        return result

    async def grant_group(self, space_key: str, group_name: str, data: dict) -> dict:
        """Grant permissions for a group in a space."""
        log.debug("grant_group called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = await self._client.post(
            f"space/{space_key}/permission/group/{group_name}", json_data=data
        )
        log.info("grant_group succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    async def grant_user(self, space_key: str, user_key: str, data: dict) -> dict:
        """Grant permissions for a user in a space."""
        log.debug("grant_user called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = await self._client.post(
            f"space/{space_key}/permission/user/{user_key}", json_data=data
        )
        log.info("grant_user succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result

    async def revoke_anonymous(self, space_key: str, data: dict) -> dict:
        """Revoke anonymous permissions for a space."""
        log.debug("revoke_anonymous called", {"space_key": space_key})
        result = await self._client.delete(
            f"space/{space_key}/permission/anonymous",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_anonymous succeeded", {"space_key": space_key})
        return result

    async def revoke_group(self, space_key: str, group_name: str, data: dict) -> dict:
        """Revoke permissions for a group in a space."""
        log.debug("revoke_group called", {
            "space_key": space_key,
            "group_name": group_name,
        })
        result = await self._client.delete(
            f"space/{space_key}/permission/group/{group_name}",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_group succeeded", {
            "space_key": space_key,
            "group_name": group_name,
        })
        return result

    async def revoke_user(self, space_key: str, user_key: str, data: dict) -> dict:
        """Revoke permissions for a user in a space."""
        log.debug("revoke_user called", {
            "space_key": space_key,
            "user_key": user_key,
        })
        result = await self._client.delete(
            f"space/{space_key}/permission/user/{user_key}",
            params={"operation": data.get("operation"), "type": data.get("type")},
        )
        log.info("revoke_user succeeded", {
            "space_key": space_key,
            "user_key": user_key,
        })
        return result
